src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java

```diff
@@ -105,17 +114,56 @@ public ProgrammingExerciseParticipationResource(ProgrammingExerciseParticipation
     @GetMapping("/programming-exercise-participations/{participationId}/latest-pending-submission")
     @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<ProgrammingSubmission> getLatestPendingSubmission(@PathVariable Long participationId) {
-        ProgrammingSubmission submission;
+        Optional<ProgrammingSubmission> submissionOpt;
         try {
-            submission = submissionService.getLatestPendingSubmission(participationId);
+            submissionOpt = submissionService.getLatestPendingSubmission(participationId);
         }
         catch (EntityNotFoundException | IllegalArgumentException ex) {
             return notFound();
         }
         catch (IllegalAccessException ex) {
             return forbidden();
         }
-        return ResponseEntity.ok(submission);
+        if (submissionOpt.isPresent()) {
+            ProgrammingSubmission submission = submissionOpt.get();
+            // Remove participation, is not needed in the response.
+            submission.setParticipation(null);
+            return ResponseEntity.ok(submission);
+        }
+        return ResponseEntity.ok(null);
+    }
+
+    /**
+     * For every student participation of a programming exercise, try to find a pending submission.
+     *
+     * @param exerciseId for which to search pending submissions.
+     * @return a Map of {[participationId]: ProgrammingSubmission | null}. Will contain an entry for every student participation of the exercise and a submission object if a pending submission exists or null if not.
+     */
+    @GetMapping("/programming-exercises/{exerciseId}/latest-pending-submissions")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Map<Long, Optional<ProgrammingSubmission>>> getLatestPendingSubmissionsByExerciseId(@PathVariable Long exerciseId) {
+        ProgrammingExercise programmingExercise;
+        try {
+            programmingExercise = programmingExerciseService.findById(exerciseId);
+        }
+        catch (EntityNotFoundException ex) {
+            return notFound();
+        }
+        if (!authCheckService.isAtLeastInstructorForExercise(programmingExercise)) {
+            return forbidden();
+        }
+        Map<Long, Optional<ProgrammingSubmission>> pendingSubmissions = submissionService.getLatestPendingSubmissionsForProgrammingExercise(exerciseId);
+        // Remove unnecessary data to make response smaller (exercise, student of participation).
+        pendingSubmissions = pendingSubmissions.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> {
+            Optional<ProgrammingSubmission> submissionOpt = entry.getValue();
```

Comment:
Could also be shortened by using the same code as suggested above

Reply:
True!